File: go/anyqmw/strategy_test.go
package anyqmw_test

import (
	"context"
	"errors"
	"testing"
	"time"

	"github.com/sns45/anyonce/go/anyonce"
	"github.com/sns45/anyonce/go/anyqmw"
	"github.com/sns45/anyq/go/core"
)

func TestStrategy(t *testing.T) {
	t.Run("REQ-Q-8: an in-flight error becomes a park for the lease remainder", func(t *testing.T) {
		inFlight := &anyqmw.InFlightError{LeaseUntil: time.Unix(1_700_000_005, 0), DelayMs: 4200}
		decision, err := anyqmw.Strategy(nil).Decide(context.Background(), core.StrategyContext{Err: inFlight, Attempt: 1, MaxAttempts: 4})
		if err != nil {
			t.Fatal(err)
		}
		if decision.Action != core.ActionPark || decision.DelayMs != 4200 {
			t.Fatalf("decision is %+v", decision)
		}
		if !inFlight.Translated() {
			t.Fatal("the strategy did not mark the error translated")
		}
	})

	t.Run("REQ-Q-8: a mismatch becomes a dead-letter with reason fingerprint-mismatch", func(t *testing.T) {
		mismatch := &anyqmw.MismatchError{Record: &anyonce.Record{Scope: "orders", Key: "m-1"}}
		decision, err := anyqmw.Strategy(nil).Decide(context.Background(), core.StrategyContext{Err: mismatch, Attempt: 1, MaxAttempts: 4})
		if err != nil {
			t.Fatal(err)
		}
		if decision.Action != core.ActionDeadLetter || decision.Reason != "fingerprint-mismatch" {
			t.Fatalf("decision is %+v", decision)
		}
	})

	t.Run("REQ-Q-8: every other error is delegated to the inner strategy", func(t *testing.T) {
		boom := errors.New("the handler failed")
		var seen core.StrategyContext
		inner := core.Custom("test-inner", func(_ context.Context, sc core.StrategyContext) (core.Decision, error) {
			seen = sc
			return core.Requeue(), nil
		})
		strategy := anyqmw.Strategy(inner)
		if strategy.Name() != anyqmw.StrategyName {
			t.Fatalf("strategy name is %q, want %q", strategy.Name(), anyqmw.StrategyName)
		}
		decision, err := strategy.Decide(context.Background(), core.StrategyContext{Err: boom, Attempt: 2, MaxAttempts: 4})
		if err != nil {
			t.Fatal(err)
		}
		if decision.Action != core.ActionRequeue {
			t.Fatalf("decision is %+v", decision)
		}
		if !errors.Is(seen.Err, boom) || seen.Attempt != 2 || seen.MaxAttempts != 4 {
			t.Fatalf("the inner strategy saw %+v", seen)
		}
	})

	t.Run("REQ-Q-8: a nil inner strategy delegates to RetryThenDeadLetter", func(t *testing.T) {
		// A retryable anyq error is what RetryThenDeadLetter retries, so the attempt cap is the only thing
		// that can turn this into a dead letter. Below the cap it retries, at the cap it dead-letters.
		retryable := core.NewConnectionError("the broker is unreachable", nil)
		strategy := anyqmw.Strategy(nil)

		retry, err := strategy.Decide(context.Background(), core.StrategyContext{Err: retryable, Attempt: 1, MaxAttempts: 4})
		if err != nil {
			t.Fatal(err)
		}
		if retry.Action != core.ActionRetry {
			t.Fatalf("below the cap the decision is %+v", retry)
		}

		dead, err := strategy.Decide(context.Background(), core.StrategyContext{Err: retryable, Attempt: 4, MaxAttempts: 4})
		if err != nil {
			t.Fatal(err)
		}
		if dead.Action != core.ActionDeadLetter || dead.Reason != "max attempts exceeded" {
			t.Fatalf("at the cap the decision is %+v", dead)
		}
	})
}

File: packages/anyq/test/strategy.test.ts
import { describe, expect, test } from 'bun:test';
import { MemoryStore } from '@anyonce/core';
import type { RetryStrategy, RetryStrategyContext } from '@anyq/core';
import { FingerprintMismatchError, InFlightError } from '../src/errors';
import { idempotent } from '../src/idempotent';
import { IDEMPOTENCY_STRATEGY_NAME, idempotencyStrategy } from '../src/strategy';
import { fakeMessage } from './fake';

function context(error: Error): RetryStrategyContext {
  return { message: fakeMessage({ body: {} }), error, attempt: 1, maxAttempts: 4 };
}

describe('idempotencyStrategy', () => {
  test('REQ-Q-8: an in-flight error becomes a park for the lease remainder', async () => {
    const strategy = idempotencyStrategy();
    const error = new InFlightError(1_700_000_005_000, 4_200);
    await expect(strategy.decide(context(error))).resolves.toEqual({
      action: 'park',
      delayMs: 4_200,
    });
    expect(error.translated).toBe(true);
  });

  test('REQ-Q-8: a mismatch becomes a dead-letter with reason fingerprint-mismatch', async () => {
    const strategy = idempotencyStrategy();
    const error = new FingerprintMismatchError({
      scope: 'orders',
      key: 'msg-1',
      fingerprint: 'aa',
      state: 'completed',
      fence: 1,
      leaseUntil: 0,
      createdAt: 0,
      expiresAt: 1,
    });
    await expect(strategy.decide(context(error))).resolves.toEqual({
      action: 'deadLetter',
      reason: 'fingerprint-mismatch',
    });
    expect(error.retryable).toBe(false);
  });

  test('REQ-Q-8: every other error is delegated to the inner strategy', async () => {
    const seen: Error[] = [];
    const inner: RetryStrategy = {
      name: 'test-inner',
      decide(ctx) {
        seen.push(ctx.error);
        return { action: 'requeue' };
      },
    };
    const strategy = idempotencyStrategy(inner);
    const other = new Error('downstream timeout');
    await expect(strategy.decide(context(other))).resolves.toEqual({ action: 'requeue' });
    expect(seen).toEqual([other]);
    expect(strategy.name).toBe(IDEMPOTENCY_STRATEGY_NAME);
  });

  test('REQ-Q-8: the default inner strategy is retryThenDeadLetter', async () => {
    const strategy = idempotencyStrategy();
    const decision = await strategy.decide({
      message: fakeMessage({ body: {} }),
      error: new Error('downstream timeout'),
      attempt: 9,
      maxAttempts: 4,
    });
    expect(decision).toEqual({ action: 'deadLetter', reason: 'max attempts exceeded' });
  });

  test('REQ-Q-8: with no strategy the typed error reaches anyq untranslated and the door warns once', async () => {
    const store = new MemoryStore();
    const warnings: string[] = [];
    const now = 1_700_000_000_000;
    const handler = idempotent(async () => {}, {
      store,
      leaseMs: 5_000,
      clock: () => now,
      logger: { warn: (message) => warnings.push(message) },
    });
    const { messageFingerprint } = await import('../src/fingerprint');
    await store.begin(
      { scope: 'orders', key: 'msg-1', fingerprint: await messageFingerprint({ a: 1 }) },
      { leaseMs: 5_000, ttlMs: 60_000, now },
    );
    for (let i = 0; i < 3; i++) {
      await handler(fakeMessage({ id: 'msg-1', body: { a: 1 } })).catch(() => {});
    }
    expect(warnings.length).toBe(1);
    expect(warnings[0]).toContain('idempotencyStrategy');
  });

  test('REQ-Q-8: a translated in-flight error does not produce the warning', async () => {
    const store = new MemoryStore();
    const warnings: string[] = [];
    const now = 1_700_000_000_000;
    const strategy = idempotencyStrategy();
    const handler = idempotent(async () => {}, {
      store,
      leaseMs: 5_000,
      clock: () => now,
      logger: { warn: (message) => warnings.push(message) },
    });
    const { messageFingerprint } = await import('../src/fingerprint');
    await store.begin(
      { scope: 'orders', key: 'msg-1', fingerprint: await messageFingerprint({ a: 1 }) },
      { leaseMs: 5_000, ttlMs: 60_000, now },
    );
    for (let i = 0; i < 3; i++) {
      const message = fakeMessage({ id: 'msg-1', body: { a: 1 } });
      const error = await handler(message).catch((e: unknown) => e);
      await strategy.decide({ message, error: error as Error, attempt: 1, maxAttempts: 4 });
    }
    expect(warnings).toEqual([]);
  });
});
